packages/react-dom/src/server/ReactPartialRenderer.js

```diff
@@ -642,8 +643,19 @@ type Frame = {
 
 type FrameDev = Frame & {
   debugElementStack: Array<ReactElement>,
+  debugReturn: ?FrameDev,
 };
 
+function pushFrame(stack: Array<Frame>, frame: Frame): void {
+  if (__DEV__) {
+    const returnFrame =
+      stack.length > 0 ? ((stack[stack.length - 1]: any): FrameDev) : null;
```

Comment:
Do we need this explicit DEV-only pointer? Isn't it always the frame at index - 1?

Reply:
Yes, but then we need to get a reference to the frame. I could do it — it's just very confusing to think about what `currentDebugStacks` would mean.

I guess my solution is also confusing.

Reply:
Not sure I follow. We only use the `debugReturn` attribute in `getStackAddendum()` where we just start at the top frame in `currentDebugFrames` and walk back. Seems like we could just do that with a `for` loop?

Reply:
No, `currentDebugFrames` is not the same thing as frames :-(

That's why this is so confusing. I rewrote and pushed a more explicit solution which doesn't try to hide that we have stacks of stacks of stacks. I hope that will clarify.

Reply:
Huh. The change you pushed to `getStackAddendum` in 5e3f51e looks like what I was suggesting 😅 but it looks like some commits are missing now so I can't step backwards....